src/shared/awsagent/roles.go

```diff
@@ -173,48 +175,185 @@ func (a *Agent) CreateRolesAnywhereProfileForRole(ctx context.Context, role type
+		for _, existingStatement := range existingPolicyStructured.Statement {
+			// Fields: Effect, Action, Resource, Principal, Sid, Condition
+			// Action is a slice
+			// Principal is a map
+			// Condition is a map
+			if newStatement.Effect != existingStatement.Effect {
+				continue
 			}
-		}
-		if !useSoftDeleteStrategy && HasSoftDeleteStrategyTagSet(role.Tags) {
-			err = a.UnsetRoleSoftDeleteStrategyTag(ctx, role)
-			if err != nil {
-				return nil, errors.Wrap(err)
+			slices.Sort(existingStatement.Action)
+			slices.Sort(newStatement.Action)
+			if !slices.Equal(existingStatement.Action, newStatement.Action) {
+				continue
+			}
+			if newStatement.Resource != existingStatement.Resource {
+				continue
+			}
+			existingPrincipals := lo.MapToSlice(existingStatement.Principal, func(key string, value string) string {
+				return fmt.Sprintf("%s:%s", key, value)
+			})
+			newPrincipals := lo.MapToSlice(newStatement.Principal, func(key string, value string) string {
+				return fmt.Sprintf("%s:%s", key, value)
+			})
+			slices.Sort(existingPrincipals)
+			slices.Sort(newPrincipals)
+			if !slices.Equal(existingPrincipals, newPrincipals) {
+				continue
+			}
+			if existingStatement.Sid != newStatement.Sid {
+				continue
+			}
+			if len(existingStatement.Condition) != len(newStatement.Condition) {
+				continue
 			}
+			existingCondition := lo.MapToSlice(existingStatement.Condition, func(key string, value any) string {
+				return fmt.Sprintf("%s:%s", key, value)
+			})
+			newCondition := lo.MapToSlice(newStatement.Condition, func(key string, value any) string {
+				return fmt.Sprintf("%s:%s", key, value)
+			})
+			slices.Sort(existingCondition)
+			slices.Sort(newCondition)
+			if !slices.Equal(existingCondition, newCondition) {
+				continue
+			}
+			found = true
+			break
+
+		}
+		if !found {
+			return false, nil
 		}
+	}
 
-		return role, nil
+	return true, nil
+}
+
+// CreateOtterizeIAMRole creates a new IAM role for service, if one doesn't exist yet
+func (a *Agent) CreateOtterizeIAMRole(ctx context.Context, namespaceName string, accountName string, useSoftDeleteStrategy bool, additionalTrustRelationshipEntries []StatementEntry) (*types.Role, error) {
+	logger := logrus.WithField("namespace", namespaceName).WithField("account", accountName)
+	exists, role, err := a.GetOtterizeRole(ctx, namespaceName, accountName)
+
+	if err != nil {
+		return nil, errors.Wrap(err)
+	}
+
+	trustPolicy := a.generateCompleteTrustPolicyAnyIdentityProvider(logger, namespaceName, accountName, additionalTrustRelationshipEntries)
+
+	identicalPolicyDocuments := false
+	if exists {
+		identicalPolicyDocuments, err = a.IsPolicyDocumentsIdenticalUnordered(role, trustPolicy)
+		if err != nil {
+			return nil, errors.Wrap(err)
+		}
 	}
 
-	trustPolicy, err := a.generateTrustPolicy(namespaceName, accountName)
+	serializedPolicyDocument, err := json.Marshal(trustPolicy)
 	if err != nil {
 		return nil, errors.Wrap(err)
 	}
 
+	if exists {
```

Comment:
move the content of this `if` block to a dedicated function `updateExistingOtterizeIAMRole`